File: src/main/java/engine/planner/PropagationPlanner.java
package engine.planner;

import engine.graph.StateNode;
import java.util.*;

public PropagationPlan<T> buildPlan(StateNode<T> start) {
    List<StateNode<T>> ordered = new ArrayList<>();
    Set<StateNode<T>> visited = new HashSet<>();

    dfs(start, visited, ordered);
    return new PropagationPlan<>(ordered);
}

File: src/main/ts/engine/planner/PropagationPlanner.ts
import { StateNode } from "../graph/StateNode"

export class PropagationPlanner<T> {
  buildPlan(start: StateNode<T>): StateNode<T>[] {
    const visited = new Set<StateNode<T>>()
    const ordered: StateNode<T>[] = []

    const dfs = (node: StateNode<T>) => {
      if (visited.has(node)) return
      visited.add(node)

      node.dependents.forEach(dep => dfs(dep))
      ordered.push(node)
    }

    dfs(start)
    return ordered
  }
}